FileUpload: uploadedFile.write(path) throws exception "org.owasp.esapi.ESAPI not found"

## 1) Environment
- PrimeFaces version: 7.0
- Application server + version: n.a. (tested with Tomcat 8.5.27, PF Test Jetty 9.4.15.v20190215)
- Affected browsers: n.a. (tested with Firefox 66, Chrome 74)

## 2) Expected behavior
Files uploaded with the `UploadedFile#write(String)` method will be written to the specified path.
This was working fine with PrimeFaces 6.2.

## 3) Actual behavior
An exception is thrown: `java.lang.NoClassDefFoundError: org/owasp/esapi/ESAPI`
<details><summary>Full stacktrace</summary>
<pre>
WARNING: /test.xhtml @18,78 fileUploadListener="#{fileBean.handleFileUpload}": java.lang.NoClassDefFoundError: org/owasp/esapi/ESAPI
javax.el.ELException: /test.xhtml @18,78 fileUploadListener="#{fileBean.handleFileUpload}": java.lang.NoClassDefFoundError: org/owasp/esapi/ESAPI
	at com.sun.faces.facelets.el.TagMethodExpression.invoke(TagMethodExpression.java:111)
	at org.primefaces.component.fileupload.FileUpload.broadcast(FileUpload.java:63)
	at javax.faces.component.UIViewRoot.broadcastEvents(UIViewRoot.java:755)
	at javax.faces.component.UIViewRoot.processDecodes(UIViewRoot.java:931)
	at com.sun.faces.lifecycle.ApplyRequestValuesPhase.execute(ApplyRequestValuesPhase.java:78)
	at com.sun.faces.lifecycle.Phase.doPhase(Phase.java:101)
	at com.sun.faces.lifecycle.LifecycleImpl.execute(LifecycleImpl.java:198)
	at javax.faces.webapp.FacesServlet.service(FacesServlet.java:658)
	at org.eclipse.jetty.servlet.ServletHolder.handle(ServletHolder.java:865)
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1623)
	at org.eclipse.jetty.websocket.server.WebSocketUpgradeFilter.doFilter(WebSocketUpgradeFilter.java:214)
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1610)
	at org.eclipse.jetty.servlet.ServletHandler.doHandle(ServletHandler.java:540)
	at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:146)
	at org.eclipse.jetty.security.SecurityHandler.handle(SecurityHandler.java:548)
	at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:132)
	at org.eclipse.jetty.server.handler.ScopedHandler.nextHandle(ScopedHandler.java:257)
	at org.eclipse.jetty.server.session.SessionHandler.doHandle(SessionHandler.java:1701)
	at org.eclipse.jetty.server.handler.ScopedHandler.nextHandle(ScopedHandler.java:255)
	at org.eclipse.jetty.server.handler.ContextHandler.doHandle(ContextHandler.java:1345)
	at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:203)
	at org.eclipse.jetty.servlet.ServletHandler.doScope(ServletHandler.java:480)
	at org.eclipse.jetty.server.session.SessionHandler.doScope(SessionHandler.java:1668)
	at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:201)
	at org.eclipse.jetty.server.handler.ContextHandler.doScope(ContextHandler.java:1247)
	at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:144)
	at org.eclipse.jetty.server.handler.ContextHandlerCollection.handle(ContextHandlerCollection.java:220)
	at org.eclipse.jetty.server.handler.HandlerCollection.handle(HandlerCollection.java:126)
	at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:132)
	at org.eclipse.jetty.server.Server.handle(Server.java:502)
	at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:370)
	at org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:267)
	at org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:305)
	at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:103)
	at org.eclipse.jetty.io.ChannelEndPoint$2.run(ChannelEndPoint.java:117)
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.runTask(EatWhatYouKill.java:333)
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.doProduce(EatWhatYouKill.java:310)
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.tryProduce(EatWhatYouKill.java:168)
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.run(EatWhatYouKill.java:126)
	at org.eclipse.jetty.util.thread.ReservedThreadExecutor$ReservedThread.run(ReservedThreadExecutor.java:366)
	at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:765)
	at org.eclipse.jetty.util.thread.QueuedThreadPool$2.run(QueuedThreadPool.java:683)
	at java.lang.Thread.run(Thread.java:748)
Caused by: java.lang.NoClassDefFoundError: org/owasp/esapi/ESAPI
	at org.primefaces.shaded.owasp.esapi.errors.EnterpriseSecurityException.<init>(EnterpriseSecurityException.java:43)
	at org.primefaces.shaded.owasp.esapi.errors.ValidationException.<init>(ValidationException.java:48)
	at org.primefaces.util.FileUploadUtils.getValidFilePath(FileUploadUtils.java:107)
	at org.primefaces.model.NativeUploadedFile.write(NativeUploadedFile.java:131)
	at org.primefaces.test.FileBean.handleFileUpload(FileBean.java:20)
	at org.primefaces.test.FileBean$$OwbNormalScopeProxy0.handleFileUpload(org/primefaces/test/FileBean.java)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.apache.el.parser.AstValue.invoke(AstValue.java:247)
	at org.apache.el.MethodExpressionImpl.invoke(MethodExpressionImpl.java:267)
	at org.apache.webbeans.el22.WrappedMethodExpression.invoke(WrappedMethodExpression.java:50)
	at com.sun.faces.facelets.el.TagMethodExpression.invoke(TagMethodExpression.java:105)
	... 42 more
Caused by: java.lang.ClassNotFoundException: org.owasp.esapi.ESAPI
	at org.codehaus.plexus.classworlds.strategy.SelfFirstStrategy.loadClass(SelfFirstStrategy.java:50)
	at org.codehaus.plexus.classworlds.realm.ClassRealm.unsynchronizedLoadClass(ClassRealm.java:271)
	at org.codehaus.plexus.classworlds.realm.ClassRealm.loadClass(ClassRealm.java:247)
	at org.codehaus.plexus.classworlds.realm.ClassRealm.loadClass(ClassRealm.java:239)
	at org.eclipse.jetty.webapp.WebAppClassLoader.loadClass(WebAppClassLoader.java:565)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:357)
	... 56 more
</pre>
</details>

## 4) Cause
When classes from org.owasp.esapi:esapi are shaded into the final PrimeFaces JAR the compiled class `org.primefaces.shaded.owasp.esapi.errors.EnterpriseSecurityException` still has an import for `org.owasp.esapi.ESAPI` (instead of a non-existant `org.primefaces.shaded.owasp.esapi.ESAPI`) because this class isn't added to the final JAR. 
Adding ESAPI as an additional dependency results in a new exception, because PrimeFaces expects the shaded logger and not the original one:
`java.lang.NoSuchMethodError: org.owasp.esapi.ESAPI.getLogger(Ljava/lang/String;)Lorg/primefaces/shaded/owasp/esapi/Logger;`

## 5) Additional thoughts
If all necessary classes were included we'd get a ValidationException from FileUploadUtils (https://github.com/primefaces/primefaces/blob/c37b93c11b81a390789e41f37d3b8feb73ff71f3/src/main/java/org/primefaces/util/FileUploadUtils.java#L106-L108) because the `getValidFilePath(...)` method checks if a file exists, meaning we can't write new uploaded files with the `write` method. The `getValidFilePath(...)` is used in both implementations of `UploadedFile`.
Our current workaroud is to get an InputStream from the UploadedFile instead and write that to disk manually.

## 6) Steps to reproduce
See test repository: https://github.com/teatwig/primefaces-test-fileupload

--- a/src/main/java/org/primefaces/model/NativeUploadedFile.java
+++ b/src/main/java/org/primefaces/model/NativeUploadedFile.java
@@ -24,6 +24,7 @@
 package org.primefaces.model;
 
 import org.primefaces.component.fileupload.FileUpload;
+import org.primefaces.shaded.owasp.SafeFile;
 
 import java.io.ByteArrayOutputStream;
 import java.io.IOException;
@@ -130,7 +131,8 @@ public class NativeUploadedFile implements UploadedFile, Serializable {
 
     @Override
     public void write(String filePath) throws Exception {
-        String validFileName = FileUploadUtils.getValidFilename(FilenameUtils.getName(filePath));
+        SafeFile file = new SafeFile(filePath);
+        String validFileName = FileUploadUtils.getValidFilename(FilenameUtils.getName(file.getPath()));
 
         if (parts != null) {
             for (int i = 0; i < parts.size(); i++) {

--- a/src/main/java/org/primefaces/model/NativeUploadedFile.java
+++ b/src/main/java/org/primefaces/model/NativeUploadedFile.java
@@ -35,6 +35,8 @@ import java.util.ArrayList;
 import java.util.List;
 import javax.faces.FacesException;
 import javax.servlet.http.Part;
+
+import org.apache.commons.io.FilenameUtils;
 import org.apache.commons.io.input.BoundedInputStream;
 import org.primefaces.util.FileUploadUtils;
 
@@ -128,16 +130,16 @@ public class NativeUploadedFile implements UploadedFile, Serializable {
 
     @Override
     public void write(String filePath) throws Exception {
-        String validFilePath = FileUploadUtils.getValidFilePath(filePath);
+        String validFileName = FileUploadUtils.getValidFilename(FilenameUtils.getName(filePath));
 
         if (parts != null) {
             for (int i = 0; i < parts.size(); i++) {
                 Part p = parts.get(i);
-                p.write(validFilePath);
+                p.write(validFileName);
             }
         }
         else {
-            part.write(validFilePath);
+            part.write(validFileName);
         }
     }
 

